src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -30,7 +30,25 @@ void GcsTaskManager::Stop() {
 
 std::vector<rpc::TaskEvents> GcsTaskManager::GcsTaskManagerStorage::GetTaskEvents()
     const {
-  return task_events_;
+  std::vector<rpc::TaskEvents> ret;
+  // NOTE(rickyx): This could be done better if we expose an iterator - which we
+  // probably have to do if we are supporting pagination in the future.
+  // As for now, this will make sure data is returned w.r.t insertion order, so we could
+  // return the more recent entries when limit applies.
+  RAY_CHECK(next_idx_to_overwrite_ == 0 || next_idx_to_overwrite_ < task_events_.size())
+      << "next_idx_to_overwrite=" << next_idx_to_overwrite_
+      << " should be in bound. (size=" << task_events_.size() << ")";
+  // Copy from the next insertion index
+  std::copy(task_events_.begin() + next_idx_to_overwrite_,
+            task_events_.end(),
+            std::back_inserter(ret));
+  // Copy the wrapped around if any
+  if (next_idx_to_overwrite_ > 0) {
```

Comment:
is this if necessary? I guess it will be just no-op?

Reply:
oh I meant we don't need if (next_idx_to_overwrite_ > 0). but it is minor lol

Reply:
oh i see. gotcha.